_posts/2019-05-06-live-view-with-presence.md

```diff
@@ -0,0 +1,581 @@
+
+```elixir
+# lib/phat_web/controllers/chat_controller.ex
+
+defmodule PhatWeb.ChatController do
+  use PhatWeb, :controller
+  alias Phat.Chats
+  alias Phoenix.LiveView
+  alias PhatWeb.ChatLiveView
+
+  def show(conn, %{"id" => chat_id}) do
+    chat = Chats.get_chat(chat_id)
+
+    LiveView.Controller.live_render(
+      conn,
+      ChatLiveView,
+      session: %{chat: chat, current_user: conn.assigns.current_user}
+    )
+  end
+end
+```
+
+The `ChatLiveView.mount/2` function sets up the initial state of the LiveView socket with the given chat, an empty message changeset with which to populate the form for a new message, and the current user:
+
+```elixir
+# lib/phat_web/live/chat_live_view.ex
+
+defmodule PhatWeb.ChatLiveView do
+  use Phoenix.LiveView
+  alias Phat.Chats
+
+  def render(assigns) do
+    PhatWeb.ChatView.render("show.html", assigns)
+  end
+
+  def mount(%{chat: chat, current_user: current_user}, socket) do
+    {:ok,
+     assign(socket,
+       chat: chat,
+       message: Chats.change_message(),
+       current_user: current_user,
+     )}
+  end
+end
+```
+
+After mounting and setting the socket state, the live view will render the `ChatView`'s `show.html` template:
+
+```elixir
+# lib/phat_web/templates/chat/show.html.leex
+
+<h2><%= @chat.room_name %></h2>
+<%=for message <- @chat.messages do %>
+  <p>
+    <%= message.user.first_name %>: <%= message.content %>
+  </p>
+<% end %>
+
+<div class="form-group">
+  <%= form_for @message, "#", [phx_submit: :message], fn _f -> %>
+    <%= text_input :message, :content, placeholder: "write your message here..." %>
+    <%= hidden_input :message, :user_id, value: @current_user.id  %>
+    <%= hidden_input :message, :chat_id, value: @chat.id  %>
+    <%= submit "submit" %>
+  <% end %>
+</div>
+```
+
+Our template is simple: it grabs the chat we assigned to our live view's socket, displays the chat room name and iterates over the messages to show us the content and sender. It also contains a form for a new message, built on the empty message changeset we assigned to our socket. At this point, our rendered template looks something like this:
+
+![]({% asset live-view-1.png @path %})
+
+### Pushing Messages to the LiveView Client
+
+Now that our live view is up and running, let's take a look at what happens when a given user submits the form for a new message.
+
+We've attached the `phx-submit` event to our form's submission, and instructed it to emit an event of type `"message"`.
+
+```elixir
+# lib/phat_web/templates/chat/show.html.leex
+
+  <%= form_for @message, "#", [phx_submit: :message], fn _f -> %>
+  ...
+```
+
+Now, we need to teach our live view how to handle this event by defining a matching `handle_event/3` function.
+
+```elixir
+# lib/phat_web/live/chat_live_view.ex
+
+defmodule PhatWeb.ChatLiveView do
+  ...
+  def handle_event("message", %{"message" => message_params}, socket) do
```

Comment:
Will this match? I am asking cause the form event is an atom and this checks for a string.

Reply:
Yep it will match. The even type gets sent as a string. I will add a note that explains that here.